### ISSUE TITLE
Performance problems of totality checker when many type parameters and recursion

### ISSUE THREAD

**Issue by buzden** [idris-lang/Idris2#1277]
I discovered that addition of a fifth type parameter to a datatype in my codebase leads to great slowdown of compilation. A code that compiled several seconds starts to compile more than several hours.

Commenting out three existing recursive calls leads to compiling withing several minutes; four commented out recursive calls leads to compiling for only several seconds.

# Steps to Reproduce

To show the problem, I tried to extract a simple example from my existing code.
It compiles for only two minutes.

```idris
module LoooongRecur

%default total

data TTy : Nat -> Nat -> Nat -> Nat -> Bool -> Type where

  TA : TTy p1 r1 i1 i2 False ->
       TTy i1 i2 i1 u2 False ->
       TTy i1 i2 p2 b1 b2    ->
       TTy p1 r1 p1 i2 False

  TB : TTy v1 r1 v2 r2 bt ->
       TTy v1 r1 v1 r2 (be && bt)

  TC : TTy p1 r1 v2 r2 False ->
       TTy v2 r2 v3 r3 b2    ->
       TTy p1 r1 v3 r3 b2

n : Nat -> Nat
n = (+ 2)

showInd : (indent : Nat) -> TTy p1 r1 v3 r3 br -> String
showInd i (TA x1 x2 x3) = if False
                            then (if False then show' i else show' i)
                            else showInd (n i) x1 ++ show' (n i)
                            where
                              show' : Nat -> String
                              show' i = showInd (n i) x3 ++ if False then "" else showInd (n i) x2
showInd i (TB x) = showInd (n i) x ++ if False then "" else showInd (n i) x
showInd i (TC x y) = if False then showInd i y else showInd i x ++ showInd i y
```

On my machine running
```
rm build/ -r && time idris2 --check LoooongRecur.idr
```
gives
```
1/1: Building LoooongRecur (LoooongRecur.idr)

real    2m3.630s
user    2m2.868s
sys     0m0.639s
```

Every thing is important: removing a type parameter or a single data constructor leads to instant compiling. Removing recursive calls, or branching, or a parameter of a data constructor leads to significant speed up.

# Expected Behavior

Fast compilation or at least independence of compilation time when somewhat small change in code.

# Observed Behavior

Very slow compilation times.

Being run with `--log 10` option, it shows that all the slow work happens in the termination checker.

**Comment by jfdm** [comment]
I have not looked at this, but adding some more information, when running Idris2 with the `--timing` flag we see:

```
$ idris2 --timing --check Long.idr
1/1: Building Long (Long.idr)
TIMING ++ Parsing Long.idr: 0.065s
TIMING ++ Reading imports: 0.031s
TIMING ++ Processing decls: 144.586s
TIMING ++ Compile defs: 0.002s
TIMING + Elaborating Long.idr: 144.686s
TIMING + Build deps: 144.864s
TIMING + Loading main file: 144.865s
```

If we make the file covering, by commenting out `%default total` we see the following improvement:

```
$ idris2 --timing --check Long.idr
1/1: Building Long (Long.idr)
TIMING ++ Parsing Long.idr: 0.065s
TIMING ++ Reading imports: 0.030s
TIMING ++ Processing decls: 0.101s
TIMING ++ Compile defs: 0.002s
TIMING + Elaborating Long.idr: 0.200s
TIMING + Build deps: 0.310s
TIMING + Loading main file: 0.311s
```

**Comment by edwinb** [comment]
I'm currently getting 13s for this, which is of course still too slow, but at least not 2m! It looks like it's spending a lot of time traversing the size change graph, which is odd because it's not that big.

My not yet pushed patch for #1365 sadly didn't have any effect. I'd be a bit surprised if any of the recent performance changes had improved it but I suppose it's possible since it seems to be all about List code. Anyway, I'll prod it some more and see if I can find a way to speed it up - it feels like this shouldn't be a challenge to the termination checker after all.

**Comment by edwinb** [comment]
Turns out it's a complex size change graph with lots of sharing that we aren't paying attention to. I added a bit to cache intermediate results and now the totality check runs instantly. I'll need to polish it a bit since my current solution is a bit hacky, but I'll fix it before releasing 0.4.0, which I'm doing very soon.
